pydis_site/apps/api/viewsets/bot/nomination.py

```diff
@@ -1,32 +1,224 @@
         """
-        DRF method for updating a Nomination.
+        DRF method for creating a Nomination.
 
         Called by the Django Rest Framework in response to the corresponding HTTP request.
         """
+        for field in request.data:
+            if field in self.frozen_on_create:
+                raise ValidationError({field: ['This field cannot be set at creation.']})
+
+        user_id = request.data.get("user")
+        if Nomination.objects.filter(active=True, user__id=user_id):
+            raise ValidationError({'active': ['There can only be one active nomination.']})
 
+        serializer = self.get_serializer(
+            data=ChainMap(
+                request.data,
+                {"active": True}
+            )
+        )
+        serializer.is_valid(raise_exception=True)
+        self.perform_create(serializer)
+        headers = self.get_success_headers(serializer.data)
+        return Response(serializer.data, status=status.HTTP_201_CREATED, headers=headers)
+
+    def partial_update(self, request, *args, **kwargs):
+        """
+        DRF method for updating a Nomination.
+
+        Called by the Django Rest Framework in response to the corresponding HTTP request.
+        """
         for field in request.data:
             if field in self.frozen_fields:
                 raise ValidationError({field: ['This field cannot be updated.']})
 
         instance = self.get_object()
+
+        if instance.active and request.data.get('unnominate_reason'):
+            raise ValidationError(
+                {'unnominate_reason': ["An active nomination can't have an unnominate reason."]}
+            )
+
+        serializer = self.get_serializer(instance, data=request.data, partial=True)
+        serializer.is_valid(raise_exception=True)
+        serializer.save()
+
+        return Response(serializer.data)
+
+    @action(detail=True, methods=['patch'])
+    def dismiss(self, request, pk=None):
+        """
+        DRF action for ending an active nomination.
+
+        Creates an API endpoint /bot/nominations/{id}/end_nomination to end a nomination. See
+        the class docstring for documentation.
+        """
+        for field in request.data:
+            if field != "unnominate_reason":
+                raise ValidationError({field: ['This field cannot be set at end_nomination.']})
+
+        if "unnominate_reason" not in request.data:
+            raise ValidationError(
+                {'unnominate_reason': ['This field is required when ending a nomination.']}
+            )
+
+        instance = self.get_object()
+        if not instance.active:
+            raise ValidationError({'active': ['A nomination must be active to be ended.']})
+
         serializer = self.get_serializer(instance, data=request.data, partial=True)
         serializer.is_valid(raise_exception=True)
+        instance.active = False
+        instance.unwatched_at = timezone.now()
         serializer.save()
+        instance.save()
```

Comment:
hm, honestly, I'm not entirely sold on the separate endpoint. why can't we do this with the standard put endpoint?

Reply:
Semantically, `PATCH` is more appropriate because `PUT` is meant to replace the entire object. However, `PATCH` is already being utilised for a more generic modification.

That being said, it does seem possible to merge functionality into just `PATCH`. The validation logic would pivot against whether the `active` field is being set.

Reply:
That's why I chose for `PATCH` instead of `PUT`, since that was my understanding as well.

I chose for a different ent point because the function of the two actions was semantically different: The regular patch to update a field of a nomination in its current state (whether it's active or inactive; it's a minor partial update); the `end` endpoint to alter the total state of the nomination. 

I noticed that the `infraction` viewset had a different endpoint for the `extended` view (`/bot/infractions/expanded`) and I liked the explicitness in that. However, infractions also change `active` to `inactive` state using the regular endpoint, so I can see the consistency point.

Reply:
I am a bit torn on this, so I will be fine whichever way we go with. @jchristgit thoughts?

Reply:
I agree with merging functionality to keep this consistent with the other endpoints, especially the infraction endpoint.

Reply:
Is this really resolved? @jchristgit suggested to merge the endpoints. I think changing it to `PUT` was a step in the wrong direction anyway because as we discussed it's less appropriate than `PATCH`

Reply:
I've moved the end nomination functionality to the PATCH endpoint now. There are already some comments on the logic, so I'm going to resolve this discussion to prevent the discussion from happening in two places.